### DISCUSSION TITLE
Wayland To Do List

### DISCUSSION THREAD

**Discussion by m-col** [qtile/qtile#2409] [General]
This is to track things that need updating or adding on the path to a stable, usable Wayland backend. I will try to keep this updated while I work on it, so this can be considered things we know are needed. If anyone posts issues about the backend when trying it out and their issue is already here, unless it needs further discussion we can probably close it in favour of following the to do list. Points are roughly ordered by priority.

NOTE: most of the main functionality has been added now :rocket:. If there is anything else that needs to be done it's probably a bug, or it's just been forgotten about and if it's the kind of thing Qtile really ought to add, consider its absence a bug too. Therefore for outstanding tasks see the issue tracker and its 'Wayland' tag: https://github.com/qtile/qtile/issues?q=is%3Aissue+is%3Aopen+label%3AWayland

- [ ] DPI, output scaling etc really needs attention
- [ ] add wlr_xdg_activation_v1
- [ ] add wlr_presentation_time
- [ ] add wlr_output_enable_adaptive_sync
- [ ] how are video calls and screen sharing?
- [ ] Partial rendering? [May have power consumption improvements.](https://mozillagfx.wordpress.com/2019/10/22/dramatically-reduced-power-usage-in-firefox-70-on-macos-with-core-animation/) will be explored during/following implementation of wlr_scene_graph (https://github.com/qtile/qtile/pull/3985)
- [x] output power management protocol https://github.com/qtile/qtile/pull/3285
- [x] idle protocol
- [x] basic libinput user configuration: https://github.com/qtile/qtile/pull/2548
- [x] Screen locking: https://github.com/qtile/qtile/pull/2852
- [x] XWayland support -- merged but bit buggy, bug reports encouraged
- [x] wlr-foreign-toplevel: #3026
- [x] Drag-and-drop
- [x] point docs at base Window class for cmd_s - https://github.com/qtile/qtile/pull/2868
- [x] Systray widget -> https://github.com/qtile/qtile/pull/2750
- [x] Tests for everything
- [x] Primary selection protocol -- when shown it is needed
- [x] Equivalent `Match` functionality -- someone who uses `Match` is better off doing this
- [x] Layer shell. Half done, requires `exclusive_zone` handling
- [x] Popups
- [x] Tree layout internal window
- [x] Z axis positioning
- [x] Drawers -> can use `shm` buffer and paint into it with cairocffi
- [x] Internal windows
- [x] Scratchpads grabbing the right window
- [x] Static windows
- [x] popup surfaces
- [x] Output damage
- [x] wlr_output_manager_v1
- [x] Window borders
- [x] wlr_virtual_keyboard_manager_v1
- [x] wlr_gamma_control_manager_v1
- [x] wlr_screencopy_manager_v1
- [x] wlr_xdg_output_manager_v1
- [x] xdg-decoration
- [x] Native wallpapers
- [x] Third party wallpapers -> is this just layer shell protocol?
- [x] window opacity
- [x] Can we make `Listener` handling and removal more managable?
- [x] Make sure IPC works as expected

Note that some things that are labelled as done may not be merged into master yet. Those that are not labelled as done have not been worked on yet and are free if anyone wants to work on them.

**Comment by flacjacket** [comment]
Are the wlr components you listed thing that we'd need to add to the pywlroots bindings? Or something that we would be doing something with in qtile?

**Reply by m-col** [reply]
Probably best added to pywlroots.

The virtual keyboard manager would probably need a class for each of `struct wlr_virtual_keyboard_manager_v1` and `struct wlr_virtual_keyboard_v1`, but they themselves are pretty small.

The screencopy and gamma_control ones are more simple, they'll just need functionality to call `wlr_screencopy_manager_v1_create` and `wlr_gamma_control_manager_v1_create`, and don't even need any objects or return values as eveything is then internal to wlroots. I _think_ `wlr_xdg_output_manager_v1_create` is the same, but IIRC that may get deprecated in favour of `wlr_output_manager_v1_create`, which will also need an object with events.

`wlr_layer_shell_v1_create` will also need an object + events, and a `struct wlr_layer_surface_v1` class most likely.

**Comment by m-col** [comment]
Pretty nice Wayland protocol explorer: https://wayland.app/protocols/

**Reply by flacjacket** [reply]
That is nice. All of the xml protocols are also scraped and included in the pywayland RTD page: https://pywayland.readthedocs.io/en/latest/module/protocol/wayland.html but the formatting and discoverability isn't quite as nice as that page.

**Comment by m-col** [comment]
About window decoration protocols, there is xdg-decoration used by some applications, and server-decoration which, although used by GTK, is becoming obselete in favour of xdg-decoration. There is a submission to GTK to migrate: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/2191

server-decoration is being removed in a future version of wlroots: https://github.com/swaywm/wlroots/blob/master/include/wlr/types/wlr_server_decoration.h#L1-L4

I've submitted implementations for xdg-decoration to pywlroots and Qtile, but I figure we can get away with not implementing server-decoration at all. Until GTK updates, windows that use it will default to using client-side decorations and draw their own title bars etc, around which our window borders are drawn.

**Reply by m-col** [reply]
Firefox also seems to use GTK to draw a title bar.

**Reply by m-col** [reply]
I think this is actually good to have in the interim, as it's been 10 months since that GTK merge request, and with little activity who knows what timescale it could be that this gets updated.

**Comment by m-col** [comment]
About clipboards/selection, I'm not sure what the situation is with them, so help is wanted if anyone wants to play around with some wayland stuff :)

Currently, without any explicit setting up of any selection/clipboard related managers, I can select text from gnome-terminal and paste it back into gnome-terminal with a middle click. I can also select, Ctrl-C, Ctrl-V to copy-paste within Firefox. However I can't paste between these clients. Some parts of the primary-selection protocol appear to be implemented in wlroots' `seat` so perhaps nothing further needs doing at all. https://github.com/swaywm/wlroots/blob/master/types/seat/wlr_seat.c

**Reply by m-col** [reply]
Looks like some clients use different protocols for this. When selecting text in kitty, I get this error:
```
[143 13:27:57.279026] [glfw error 65544]: Wayland: Cannot copy no primary selection device available
```
So I guess something that is needed is not yet implemented.

**Reply by m-col** [reply]
I added a handler for the primary selection protocol which has nicely extended the number of clients that can copy and paste between each other.

**Comment by Fabien-Bernard** [comment]
I don't see anything about restarting qtile in the todo list. As I just experienced a crash (Runtime Error : Failed to create wlroots backend), I am not sure if I should already report it.
Additional question : is it already possible to change keyboard layout?

**Reply by m-col** [reply]
Restarting doesn't really make sense when using the Wayland backend. As an X WM, the connection to the X server lets clients stay alive in the absence of the WM, which can start up again and continue its role as if nothing happened, but in Wayland Qtile _is_ the server connected to the clients, so when it exits the clients disconnect, and most will exit themselves. There is no away around this, so when somebody updates Qtile they must close their windows, exit the compositor, then start it up again.

The other reason people restart is to reload their configs. I started working on a way to reload the config without restarting in https://github.com/qtile/qtile/pull/2369 before getting carried away with the Wayland stuff. I'm now running under Wayland full-time so this does get annoying quite often, so I want to continue that work soon but I'm still busy with refactoring the tests so I can test the new backend. As for the Runtime error... I should probably just disable restarting when using this backend.

> Additional question : is it already possible to change keyboard layout?

I'm not actually sure, it isn't something I've thought about because I only use one layout! Please could you post an issue about it? This way we can keep track of it. Sway seems to handle it which might be a good place to start: https://github.com/swaywm/sway/blob/7c74f01f0ae9d5b3f92d3e6fc64cb9abe95b4c52/sway/commands/input/xkb_switch_layout.c

**Reply by Fabien-Bernard** [reply]
I am so used to hit the restart key after a change in my config file, I will add a backend condition for this key in my config to avoid launching it by inadvertence. Thank you really much for the explanation - and for all the good work on this!

I exported the XKB_DEFAULT_LAYOUT variable before launching qtile to solve my keyboard headache. I'll continue to explore the xkbcommon library before creating an issue.

**Reply by ramnes** [reply]
Yeah, just sneaking in to emphasize the "thank you", because this is a tremendous work that you've been doing by both implementing Wayland support and the first non-X Qtile core, which could open the door for some really fun stuff. Thanks a million!

**Reply by m-col** [reply]
Thanks :D

**Comment by Magnitudexx** [comment]
Is XWayland support  not a priority until people show interests? I'm looking into Wayland for a few reasons but few programs I have installed works except for my terminal emulator.

**Reply by m-col** [reply]
Pretty much... and ideally we wouldn't need it at all, as it introduces many more considerations and things that can go wrong. XWayland is largely a workaround to help people move over to Wayland, which is great, but it's just that... a workaround. I suggest looking for alternative programs that support Wayland directly -- for a terminal emulator I can suggest `foot` or `alacritty` which both work great natively.

**Reply by jjsullivan5196** [reply]
@m-col @Magnitudexx I think support for XWayland in qtile itself can be safely elided altogether, as the [cage compositor](https://github.com/Hjdskes/cage) can already nest itself in an existing wayland session and provide an XWayland server on its own. Multiple cages can also be run at the same time, if you need more than one Xorg app running at once. Here's a screenshot of chromium and xorg firefox running side-by-side on alpine linux with wayland qtile. 

![image](https://user-images.githubusercontent.com/3991905/131233859-6a9b1d7a-1041-4545-bc6c-21491fc6a550.png) 

It looks like the context menus in chromium don't work for some reason, but this is just the first time I'm trying it so it might be a dumb problem, but everything else works! Seems like the only hard 'downside' is that cage will eat any child windows of the contained app, however I actually prefer this since it removes any complexities of handling xorg client requests. I want to give this a try with steam since games are probably the highest quantity of xorg apps still in my life, I'll report back if I find anything interesting in this direction :^) This probably will cover other legacy xorg apps pretty well

**Reply by m-col** [reply]
Huh, cool. Do you get the correct context menu behaviour when running cage directly?

**Reply by jjsullivan5196** [reply]
@m-col Just tried it, yeah it works with cage by itself, maybe a protocol bug? Everything seems to be OK in firefox, so can't really say for sure.

Looks like its generally an issue with any popup-style menus, the main menu doesn't work either when nested. I might hammer on this a little harder since I want to be able to test software in chrome :^)

**Reply by m-col** [reply]
This is actually really useful. I'm using it for discord, which seems to be working 100% perfectly as if it was just natively running within Qtile. Awesome idea. Chromium has always played against the rules with it comes to X so it's possible the problems lie just with that unfortunately.

**Reply by goduck777** [reply]
Hi. My workflow relies on using SSH X11 forwarding, which works well in other wayland compositor but not in Qtile. I wonder if there is a workaround for this, as I don't want to launch my terminal in cage.

**Reply by m-col** [reply]
@goduck777 for that we would need XWayland support. We do not currently implement this I'm afraid, but an implementation would be welcome if a contributor is interested.

**Comment by Magnitudexx** [comment]
I think I have something that could be added to the To Do List if i didn't miss it. As I understand it Wayland compositors are the one handling things like dpi/scaling. Could not find a setting or config option for this but maybe i just missed it or have misunderstood something.

**Reply by m-col** [reply]
Hey @Magnitudexx, I've submitted a PR for this here: https://github.com/qtile/qtile/pull/2783. If you are able to confirm that it's working then that'd be appreciated :)

**Comment by Graeme22** [comment]
Another feature that would be nice: input inhibitor for screenlockers like waylock and swaylock. I think that's the job of the compositor to implement? For example when I try to run swaylock I get this:

```
2021-09-30 11:14:33 - [swaylock-1.5/main.c:1186] Compositor does not support the input inhibitor protocol, refusing to run insecurely
```

**Reply by m-col** [reply]
Good idea, I'll add it to the list!

**Comment by m-col** [comment]
https://github.com/qtile/qtile/pull/2360 has been merged which means wayland users can now reload their config like they could in X11.

**Reply by m-col** [reply]
To correct myself, not so much "like they could" - you need to use `lazy.reload_config` rather than `lazy.restart`.

**Comment by m-col** [comment]
Lots of people have expressed interest in XWayland, and while I have been quite hesitant to include it in Qtile, I would help out with code reviews etc if anybody is interested in taking a go at implementing XWayland support.

**Comment by anriha** [comment]
Any plans to add support for idle protocol? If not I may look into it.

**Reply by m-col** [reply]
Not yet, contribution very welcome :smile:

**Comment by raven2cz** [comment]
I see that DPI, output scaling etc is already checked. Can I test it? What has to be configured for dpi change?

**Reply by m-col** [reply]
You need to configure scaling for your output using a tool such as `kanshi`, `wlr-randr` or `wdisplays`.

**Comment by jwijenbergh** [comment]
What are your opinions on wlr_scene @m-col? AFAICT it would be a good fit for qtile, it still has some rough edges such as not implementing scissors though. Most compositors are moving towards this such as sway (has a WIP Pr). This would give us full damage tracking for free. Additionally it would simplify the codebase a lot

**Reply by m-col** [reply]
I haven't looked into it for a while, but from what I understand, Qtile is exactly the kind of compositor that would benefit from it, and I'm all for it. I suspect it would also make it much more easy to get atomic state changes WRT rendering which would be amazing. Are you interested in working on it? :wink:

**Comment by cr1ogen** [comment]
Xwayland in Gnome supports more resolutions.

In gnome I use the monitor at 4k scaled at 200% and in Qtile using wlr-randr.

If I write to a console in Gnome the xrandr command, it gives me the following output
```Screen 0: minimum 16 x 16, current 3840 x 2160, maximum 32767 x 32767
XWAYLAND0 connected primary 3840x2160+0+0 (normal left inverted right x axis y axis) 620mm x 340mm
   3840x2160     59.98*+
   2048x1536     59.95  
   1920x1440     59.90  
   1600x1200     59.87  
   1440x1080     59.99  
   1400x1050     59.98  
   1280x1024     59.89  
   1280x960      59.94  
   1152x864      59.96  
   1024x768      59.92  
   800x600       59.86  
   640x480       59.38  
   320x240       59.52  
   2560x1600     59.94  
   1920x1200     59.88  
   1680x1050     59.95  
   1440x900      59.89  
   1280x800      59.81  
   720x480       59.71  
   640x400       59.95  
   320x200       58.96  
   3200x1800     59.96  
   2880x1620     59.96  
   2560x1440     59.96  
   2048x1152     59.90  
   1920x1080     59.96  
   1600x900      59.95  
   1368x768      59.88  
   1280x720      59.86  
   1024x576      59.90  
   864x486       59.92  
   720x400       59.55  
   640x350       59.77 
 ```

and if I write it to Qtile, only this

```Screen 0: minimum 16 x 16, current 1920 x 1080, maximum 32767 x 32767
XWAYLAND0 connected 1920x1080+0+0 (normal left inverted right x axis y axis) 620mm x 340mm
   1920x1080     59.96*+
```
This makes steam games in qtile the maximum resolution is 1920x1080, while in gnome games can use resolutions up to 4k (my monitor is 4k), could something be done so that this can be implemented in qtile? Thank you

**Reply by m-col** [reply]
Definitely! That's pretty interesting, I didn't notice (but can reproduce) that XWayland isn't aware of all modes available by the outputs. I've generated an issue from this post: https://github.com/qtile/qtile/issues/3847

**Comment by jiiiijiij** [comment]
Sorry I don't know what bug should I call this (or maybe it was my setup problem): 
- Davinci Resolve workspace wont open after the project chooser menu, 
- Wine apps such as LoL game client only appear part of it inside a small float window that un-resizable,
- Also often time the login or search promt in browser (chromium base and mozilla) will cause the browser window to freeze, can't close the window too. The search box in qtile docs is one example. I think is some javascript problem, never happen with normal text box.

**Reply by m-col** [reply]
Are these all running with XWayland? There are a number of problems with that at the minute unfortunately. For firefox you can make it use wayland with `MOZ_ENABLE_WAYLAND=1` if you aren't using that already. It may fix the issue for that.

**Comment by arjan-s** [comment]
Is display color calibration possible under Wayland? I'd like to use displaycal/colord or something alike, with my ColorMunki Smile, to fix the extremely bad colors of my laptop display.

**Reply by m-col** [reply]
You might be interested in this sway thread: https://github.com/swaywm/sway/issues/1486. Tldr: it is technically possible but not something that is supported by sway or wlroots (yet). It looks like someone got it working a bit with XWayland clients specifically (if I understood correctly) though that doesn't really help you.

**Comment by Graeme22** [comment]
The naive screen locking protocol we were using previously is no longer supported by libraries like `swaylock`, which now implement the much more secure `ext-session-lock-v1` protocol. So I think the "screen locking" box should be unchecked--my fix from 3 years ago is no longer sufficient today (#2852). I may look into getting a new PR out if I have time.

Edit: This is already in `wlroots`, so just the implementation is needed!

**Reply by jwijenbergh** [reply]
Just an update from my side, I am rewriting the wayland backend in c, and have most things working. I will also implement new protocols such as ext-session-lock-v1. The reason being is that it's too hard to follow upstream pywlroots with bindings

**Reply by jwijenbergh** [reply]
also for ext-session-lock-v1, there is a WIP PR https://github.com/qtile/qtile/pull/5016

**Comment by v3ss0n** [comment]
Also please look into drag and drop (for chrome/chromium based engines) . It is broken because it hit CPU 100%
https://github.com/qtile/qtile/issues/4992

**Comment by TheHighGroundWins** [comment]
Implementing `text-input-v3` and `input-method-v2` wayland protocols would allow IME multilingual input to work on wayland. Currently only non IME languages like english, spanish, mongolian work on wayland. Languages like Japanese, Chinese are untypable on wayland natively unless its xwayland, GTK or QT programs.

Once the C rewrite is done, I'll try to implement it myself but I would have to figure out how to place and edit text on qtile's wayland backend.

**Comment by richcarni** [comment]
Hi @jwijenbergh - just wondering how development on the wayland c backend is coming along? 

Is this something where more man power could help? Putting my hand up if that would be of any use

**Comment by tych0** [comment]
I'm going to go ahead and close this thread. The last post is from @richcarni offering to help, who is now a maintainer as a result of landing the new wayc backend.
